parser/msg_parser.c

```diff
@@ -799,9 +798,9 @@ int parse_msg(char* buf, unsigned int len, struct sip_msg* msg)
 	return 0;
 
 error:
-	/* more debugging, msg->orig is/should be null terminated*/
-	LM_ERR("message=<%.*s>\n", (int)len, ZSW(buf));
-	return -1;
+		/* more debugging, msg->orig is/should be null terminated */
+		LM_ERR("message=<%.*s>\n", (int)len, ZSRW(buf));
```

Comment:
So this is a tough one. We don't want to log customer information, but if I'm understanding this whole implementation correctly what we are doing is replacing any potentially sensitive information with `*`s. The problem here is that we're now logging useless information. In the case of many of the changed logs, the log of the error *could* still be helpful even without the info? And in most cases the redacted info is small. In this case, we will be logging and entire SIP message as just `*`s, which seems excessive.
It's necessary in this implementation to keep the log call in place in case the compiler define is disabled, but that makes me wonder if this is the correct approach. Given the relatively small number of places changes so far, would it be better to wrap the log itself in the define, rather than log useless information?

Reply:
My first thought on the issue is that we should implement a general log redact module where the user can define regex patterns to replace. This way, it would be entirely up to the users to decide what counts as PII, and a general solution like this may hit upstream as well. It also could solve the above case correctly by redacting certain parts of the SIP message if found.

My concern about this is the performance cost, and it looks overkill to solve the problem. Running multiple regex search & replace every time we log something does not sound right. Maybe this should be a third-party software between logging and log collecting.

Regarding the above specifically: I think it has value to show a fully redacted log because it shows that an error happened, in which function, line. It depends on the context, but if for an error only that specific log would be logged and we remove it, we will even hide the error existence, which is not good.

Reply:
I don't think regexes are a complete solution, and as you said they are expensive. SIP is a complex protocol and it's not possible to craft regexes that would capture all potential sensitive information in the logs without context of what data was being logged. I think this is the case whether the masking is done in OpenSIPS or by an external solution, which was the OpenSIPS team's suggestion.
I do think there is some value in having the redacted log for most of these logs. My primary concern, and the reason for the placement of the comment, was this log specifically, which will log an entire redacted SIP message. That seems a bit excessive. But if it is necessary for the solution, that is a sufficient answer. I only thought to spur some thought on whether there might be another solution.
Personally, I had wondered whether this could be done as a global or module variable in OpenSIPS, and configured at runtime. Specifying at compile time has issues, even for us, as it means the redaction will be occurring even in our IT tests, as well as DCA and TCA, which is not ideal. Doing a runtime configuration would require a more specific check in the code, and not this simple macro switch. But I think that may be justified.
Even something like providing a second set of logging functions, so that the code could call the "sensitive" log when it knows it is logging sensitive data. That logging function could then be responsible for checking if redaction is needed. It is more code work and requires more maintenance with new code/features, but it seems like a more robust solution.